web/src/app/(main)/games/page.tsx

```diff
@@ -70,60 +98,117 @@ export default function Games() {
-                <Box
-                    sx={{
-                        flexGrow: 1,
-                    }}
-                />
-                {loggedIn && (
-                    <div>
-                        <Button href="/games/new" LinkComponent={Link}>
-                            Create a new game
+                
+                <Box sx={{ display: 'flex', alignItems: 'center', flexGrow: 1, maxWidth: '50%' }}>
+                    <TextField
+                        label="Search games"
+                        variant="outlined"
+                        size="small"
+                        fullWidth
+                        value={searchString}
+                        onChange={(e) => updateSearchString(e.target.value)}
+                    />
+                    {searchString && (
+                        <Button
+                            onClick={() => updateSearchString('')}
+                            size="small"
+                            sx={{ ml: 1, whiteSpace: 'nowrap' }}
+                        >
+                            Clear
                         </Button>
-                    </div>
+                    )}
+                </Box>
+                
+                <Box sx={{ flexGrow: 1 }} />
+                
+                {loggedIn && (
+                    <Button href="/games/new" LinkComponent={Link}>
+                        Create a new game
+                    </Button>
                 )}
             </Box>
             {Object.keys(games)
                 .filter((k) => games[k].length > 0)
-                .map((key) => (
-                    <Box key={key}>
-                        <Typography
-                            variant="h5"
-                            sx={{
-                                pb: 1,
-                            }}
-                        >
-                            {key}
-                        </Typography>
-                        <Masonry
-                            columns={{
-                                xs: 1,
-                                sm: 2,
-                                md: 4,
-                                lg: 5,
-                                xl: 6,
-                            }}
-                            spacing={2}
-                        >
-                            {games[key].map((game, index) => (
-                                <GameCard
-                                    key={game.slug}
-                                    game={game}
-                                    index={index}
-                                    localFavorites={localFavorites}
-                                    setLocalFavorites={setLocalFavorites}
-                                />
-                            ))}
-                        </Masonry>
-                    </Box>
-                ))}
+                .map((key) => {
+                    const filteredGames = games[key].filter((game) =>
```

Comment:
Filter the original games list in the main logic of the component and then feed the filtered list into the reducer, to avoid filtering the list for each key (and it will significantly simplify the render logic)